### PR TITLE
Refactor and fixes multigpu support

### PR DESCRIPTION
**Pull request by sklam**
* Refactor the multigpu and cuda context stack management (in numba.cuda.cudadrv.devices)
    * enforces **one context per device per process** rule as introduced in CUDA 4.0 Runtime API
    * all cuda users (including tests) must use this to create context
* Fixes context leaking problems on python2.7 and python3.3 
* Fixes exception leaking in threads (numba/cuda/tests/cudadrv/test_reset_device.py)
* All cuda tests must use the new cuda context management

### CODE REVIEW

File: numba/cuda/cudadrv/devices.py [hunk 1]
```diff
@@ -1,134 +1,228 @@
...
+        ctx = self._devctx.get(gpu.id)
+        if ctx is None:
+            ctx = self._create_context(gpu)
+        else:
+            # Push the context to the stack
+            ctx.push()
+
+        return ctx
+
+    def push_context(self, gpu):
```
Comment (gmarkall) [inline numba/cuda/cudadrv/devices.py]: It looks like `push_context` will always push a context for the given GPU on the stack regardless of whether it is the current context - is that supposed to happen, or am I misunderstanding something?

File: numba/cuda/cudadrv/devices.py [hunk 2]
```diff
@@ -1,134 +1,228 @@
...
+        # integrity check
+        assert driver.get_context().value == top.handle.value, (
+            "An unmanaged CUDA context is active."
+        )
+        return top
+
+    def _create_context(self, gpu):
+        """Create a new context for the given gpu
+        """
+        # Use lock to prevent internal state mutation
```
Comment (gmarkall) [inline numba/cuda/cudadrv/devices.py]: Internal state corruption, rather than mutation?

**Comment by gmarkall** [comment]
The following:

```python
from numba import cuda
len(cuda.gpus)
```

results in:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gmarkall/work/numba/numba/cuda/cudadrv/devices.py", line 44, in __len__
    return len(self.lst)
```

- should this be expected to work?

**Comment by gmarkall** [comment]
Something I've observed related to `push_context` always pushing - if I run the following after a context for a GPU is already created:

```python
with cuda.gpus[0]:
    with cuda.gpus[0]:
        pass
```

Then during execution, there are three contexts on the stack:

```
[<CUDA context c_void_p(28950752) of device 0>,
 <CUDA context c_void_p(28950752) of device 0>,
 <CUDA context c_void_p(28950752) of device 0>]
```

It seems to me to be desirable to be able to have nested contexts in this way, since I wouldn't necessarily want to bother making sure I always avoided nesting contexts in a large program. Therefore, I think `push_context` just has an inaccurate docstring "Push a context for the given GPU if it is not the current context") rather than the implementation being incorrect - furthermore, I expect it would cause problems in `pop_context` if `push_context` did not always push a context when called. 

I've not yet spotted a case where I think this behaviour is actually going to be a problem - do you agree with this?

**Comment by gmarkall** [comment]
> Then during execution, there are three contexts on the stack:

Just to clarify - I mean the same context is pushed three times.

**Comment by sklam** [comment]
@gmarkall, It is expected to always push a context to the context stack even if it is the currently active context.  The behavior (pushing the same context 3 times) is expected in your code snippet.

**Comment by sklam** [comment]
and, @gmarkall, thanks for reviewing!

**Pull request merged**